chain/index/events.go

```diff
@@ -236,48 +238,98 @@ func loadExecutedMessages(ctx context.Context, cs ChainStore, recomputeTipSetSta
 	return ems, nil
 }
 
-// checkTipsetIndexedStatus verifies if a specific tipset is indexed based on the EventFilter.
-// It returns nil if the tipset is indexed, ErrNotFound if it's not indexed or not specified,
-func (si *SqliteIndexer) checkTipsetIndexedStatus(ctx context.Context, f *EventFilter) error {
-	var tipsetKeyCid []byte
-	var err error
-
-	// Determine the tipset to check based on the filter
-	switch {
-	case f.TipsetCid != cid.Undef:
-		tipsetKeyCid = f.TipsetCid.Bytes()
-	case f.MinHeight >= 0 && f.MinHeight == f.MaxHeight:
-		tipsetKeyCid, err = si.getTipsetKeyCidByHeight(ctx, f.MinHeight)
-		if err != nil {
-			if err == ErrNotFound {
-				// this means that this is a null round and there exist no events for this epoch
-				return nil
-			}
+// checkRangeIndexedStatus verifies if a range of heights is indexed.
+// It checks for the existence of non-null rounds at the range boundaries.
+func (si *SqliteIndexer) checkRangeIndexedStatus(ctx context.Context, f *EventFilter) error {
+	minHeight := f.MinHeight
+	maxHeight := f.MaxHeight
 
-			return xerrors.Errorf("failed to get tipset key cid by height: %w", err)
+	// Find the first non-null round in the range
+	startCid, err := si.findFirstNonNullRound(ctx, &minHeight, maxHeight)
+	if err != nil {
+		return xerrors.Errorf("failed to find first non-null round: %w", err)
+	}
+
+	// If all rounds are null, consider the range valid
+	if startCid == nil {
+		return nil
+	}
+
+	// Find the last non-null round in the range
+	endCid, err := si.findLastNonNullRound(ctx, &maxHeight, minHeight)
+	if err != nil {
+		if errors.Is(err, ErrRangeInFuture) {
+			return xerrors.Errorf("range end is in the future: %w", err)
 		}
-	default:
-		// This function distinguishes between two scenarios:
-		// 1. Missing events: The requested tipset is not present in the Index (an error condition).
-		// 2. Valid case: The tipset exists but contains no events (a normal situation).
-		// Currently, this distinction is only made for the common use case where a user requests events for a single tipset.
-		// TODO: Implement this functionality for a range of tipsets. This is expensive and not a common use case so it's deferred for now.
+		return xerrors.Errorf("failed to find last non-null round: %w", err)
+	}
+
+	// If all rounds are null, consider the range valid
+	if endCid == nil {
 		return nil
 	}
 
-	// If we couldn't determine a specific tipset, return ErrNotFound
-	if tipsetKeyCid == nil {
-		return ErrNotFound
+	// Check indexing for start and end tipsets
+	if err := si.checkTipsetByKeyCid(ctx, startCid, minHeight); err != nil {
+		return err
```

Comment:
let's think through what the error is going to look like to the user here if they started with a null round height:

* Request with minHeight `X`
* Error says: `failed to check if tipset at height X+1 is indexed`

It's a bit awkward, but you could do a check for `minHeight != f.MinHeight` and in that case return a custom error wrapping the one you got, something like `xerrors.Errorf("checking first non-null tipset after %d: %w", f.MinHeight, err)`

Reply:
same with the `endCid` case too

Reply:
actually, now I think it through more, let's not bother, the error case should be rare I think, so not worth the extra code